### PR TITLE
Allow root zone queries and fall through

### LINKED ISSUE #35: Resolving zone domain and proper fallthrough
Let's take this example
```
k8s_gateway domain.tld {
    kubeconfig /path/kubeconfig
    fallthrough
}
```
If querying for `domain.tld` the current implementation will call `serveApex` in https://github.com/ori-edge/k8s_gateway/blob/master/gateway.go#L123 and return `SOA` or `NS` entries. If `fallthrough` is declared it will won't handle it here and still return those two possible responses (inconsistent with other plugins).

I built it skipping this case and handle the zone domain same as other subdomains, so having an Ingress defining `domain.tld` in its hosts works fine and also the default `fallthrough` handler. I'm not sure why the response handler has an override for the zone domain. If `SOA` and `NS` are required we could let pass any other request to the default handler to be able to resolve the zone domain or if it's not desired at least let the zone domain fall through.

Any ideas? I'm happy to PR either way, but would prefer it could resolve the zone domain itself using k8s resources.

### PR DESCRIPTION
**Pull request by cubic3d**
This PR allows to query the zone root domain and fall through if it doesn't match a resource.

**Tests needs to be adjusted after agreement on actual behaviour.**

Fixes #35

### CODE REVIEW

File: gateway.go [hunk 1]
```diff
@@ -142,39 +143,44 @@ func (gw *Gateway) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
...
 		if gw.Fall.Through(qname) {
 			return plugin.NextOrFailure(gw.Name(), gw.Next, ctx, w, r)
 		}
-
-		m.Rcode = dns.RcodeNameError
-		m.Ns = []dns.RR{gw.soa(state)}
-		if err := w.WriteMsg(m); err != nil {
-			log.Errorf("Failed to send a response: %s", err)
-		}
-		return 0, nil
```
Comment (networkop) [inline gateway.go:159]: what happens if fallthrough is not configured but `len(addrs)` is 0?

**Review by networkop — COMMENTED** [review]
I've left a few comments.  At the very least I'm expecting this change to not affect the existing behavior, so existing tests should all pass.

File: gateway.go [hunk 2]
```diff
@@ -118,10 +118,11 @@ func (gw *Gateway) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
 		return dns.RcodeServerFailure, plugin.Error(thisPlugin, fmt.Errorf("Could not sync required resources"))
 	}
 
+	var is_root_zone_query bool
```
[the remark this answers is not in the document]
Comment (cubic3d) [reply gateway.go:121]: Worked on Python prior to that, no idea how that slipped through ;) will be fixed

File: gateway.go [hunk 3]
```diff
@@ -118,10 +118,11 @@ func (gw *Gateway) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
...
 	for _, z := range gw.Zones {
 		if state.Name() == z { // apex query
-			ret, err := gw.serveApex(state)
-			return ret, err
```
[the remark this answers is not in the document]
Comment (cubic3d) [reply gateway.go:124]: It's handled here: https://github.com/ori-edge/k8s_gateway/blob/df76f611f3ea464230cc617a2019e5f881340a09/gateway.go#L168-L182

File: gateway.go (hunk 1 above)
Comment (cubic3d) [reply gateway.go:159]: It will still return NXDOMAIN https://github.com/ori-edge/k8s_gateway/blob/df76f611f3ea464230cc617a2019e5f881340a09/gateway.go#L158-L161

**Comment by networkop** [comment]
@cubic3d thanks for the PR, can you add some tests to this ? It'd be easier to see the corner cases.

**Comment by cubic3d** [comment]
Sure, as soon as time allows ;)

**Comment by networkop** [comment]
@cubic3d check out this commit https://github.com/ori-edge/k8s_gateway/commit/5966d8d6bfa9dcb6a1093b953dfe27e316f3ac8b
I've made some changes to make your PR pass the existing tests. I for some reason couldn't push into this PR so created another branch instead. 

So now you just need to add a set of tests to cover the fallthrough behaviour

**Comment by cubic3d** [comment]
Thank you, strange that you can't push - edits are allowed. I'll get back to it at weekend.

**Comment by networkop** [comment]
thanks, will check it out later this week

**Comment by networkop** [comment]
LGTM. how does it look to you @cubic3d ?

**Comment by cubic3d** [comment]
LGTM too! Thank you! I'll mark ready
